### PR TITLE
Added KF with sparse sites and related test

### PR DESCRIPTION
**Pull request by prakharverma**
Added Kalman filter with sparse sites and related tests.

### CODE REVIEW

File: markovflow/kalman_filter.py [hunk 1]
```diff
@@ -495,3 +494,120 @@ def _log_det_observation_precision(self):
...
+    and :class:`~markovflow.emission_model.EmissionModel`, with Gaussian sites, over a time grid.
+    """
+
+    def __init__(self, state_space_model: StateSpaceModel, emission_model: EmissionModel, sites: GaussianSites,
+                 time_grid: tf.Tensor, observations_idx: tf.Tensor, observations: tf.Tensor):
+        """
+        :param state_space_model: Parameterises the latent chain.
+        :param emission_model: Maps the latent chain to the observations.
+        :param sites: Gaussian sites over the time grid.
+        :param time_grid: Time grid with shape (T,).
```
Comment (vincentadam87) [inline markovflow/kalman_filter.py]: I know I suggested it but it seems you are only ever using the shape of the time grid, so maybe you can pass just the shape (I know I suggested the full thing sorry)
this is different from a gp model where you need the actual time values to compute the covariances or transition matrices etc

File: tests/integration/test_kalman_filter_with_sparse_sites.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+from markovflow.mean_function import LinearMeanFunction
+from markovflow.models.gaussian_process_regression import GaussianProcessRegression
+from markovflow.kalman_filter import KalmanFilterWithSparseSites, UnivariateGaussianSitesNat
+from markovflow.likelihoods import MultivariateGaussian
+
+
+@pytest.fixture(name="kalman_gpr_setup")
+def _setup(batch_shape):
+    """
+    FIXME: Currently batch_shape isn't used.
```
Comment (vincentadam87) [inline tests/integration/test_kalman_filter_with_sparse_sites.py:24]:
```suggestion
    Create a Gaussian Process model and an equivalent kalman filter model
    with more latent states than observations. 
    FIXME: Currently batch_shape isn't used.
```

File: markovflow/kalman_filter.py [hunk 3]
```diff
@@ -495,3 +494,120 @@ def _log_det_observation_precision(self):
...
+        :param observations_idx: Index of the observations in the time grid with shape (N,).
+        :param observations: Observations with shape (N, state_dim).
+        """
+        self.sites = sites
+        self.time_grid = tf.expand_dims(time_grid, axis=-1)
+        self.observations_idx = observations_idx
```
Comment (vincentadam87) [inline markovflow/kalman_filter.py]: I'd prefer index to idx
I think in mf we sometimes shorten, but never abbreviate. 
but I can be proven wrong

File: markovflow/kalman_filter.py [hunk 4]
```diff
@@ -495,3 +494,120 @@ def _log_det_observation_precision(self):
...
+        return tf.scatter_nd(self.observations_idx , data_sites_precision, self.time_grid[..., None].shape)
+
+    @property
+    def _log_det_observation_precision(self):
+        """
+        Sum of log determinant of the precisions of the observation model. It only calculates for the data_sites as
+        other sites precision is anyways zero.
+        """
+        data_r_inv = self._r_inv_data
+        return tf.reduce_sum(tf.linalg.logdet(data_r_inv), axis=-1)
```
Comment (vincentadam87) [inline markovflow/kalman_filter.py]:
```suggestion
        return tf.reduce_sum(tf.linalg.logdet(self.data_sites.precisions), axis=-1)
```
I don't think we need that r_inv_data

Comment (vincentadam87) [reply markovflow/kalman_filter.py]: why not, it is actually used twice

File: tests/integration/test_kalman_filter_with_sparse_sites.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    time_grid = np.arange(0.0, 1.0, 0.01)
```
Comment (vincentadam87) [inline tests/integration/test_kalman_filter_with_sparse_sites.py]: can you do an inhomogenous time grid?
Do tests still pass?
```suggestion
    dt = 0.01
    time_grid = np.arange(0.0, 1.0, dt)
    time_grid += np.random.rand(*time_grid.shape) * dt / 2. 
```

File: tests/integration/test_kalman_filter_with_sparse_sites.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    time_points = time_grid[::10]
+    observations = np.sin(12 * time_points[..., None]) + np.random.randn(len(time_points), 1) * 0.1
+
+    input_data = (
+        tf.constant(time_points, dtype=default_float()),
+        tf.constant(observations, dtype=default_float()),
+    )
+
+    observation_covariance = 1.0  # Same as GPFlow default
+    kernel = Matern12(lengthscale=1.0, variance=1.0, output_dim=observations.shape[-1])
```
Comment (vincentadam87) [inline tests/integration/test_kalman_filter_with_sparse_sites.py:41]: could you try to have some state offset into the kernel?
That will test whether the Kf class deals well with offsets in the ssm (I m not 100% sure it will)
and we need this with our linearized SDEs

**Review by vincentadam87 — COMMENTED** [review]
First batch of comments

File: markovflow/kalman_filter.py (hunk 1 above)
Comment (prakharverma) [reply markovflow/kalman_filter.py]: Makes sense, I also didn't notice that I only use the shape.

File: markovflow/kalman_filter.py (hunk 4 above)
Comment (prakharverma) [reply markovflow/kalman_filter.py]: It's a function that does `self.data_sites.precisions` but yes that variable wasn't required so removed.

File: tests/integration/test_kalman_filter_with_sparse_sites.py (hunk 5 above)
Comment (prakharverma) [reply tests/integration/test_kalman_filter_with_sparse_sites.py]: It works! Added test cases for that too.

Comment (prakharverma) [reply tests/integration/test_kalman_filter_with_sparse_sites.py]: I just coded it up a bit differently.

File: tests/integration/test_kalman_filter_with_sparse_sites.py (hunk 6 above)
Comment (prakharverma) [reply tests/integration/test_kalman_filter_with_sparse_sites.py:41]: It does! I added it to the test script.

File: markovflow/kalman_filter.py [hunk 7]
```diff
@@ -495,3 +494,119 @@ def _log_det_observation_precision(self):
...
+
+    def __init__(self, state_space_model: StateSpaceModel, emission_model: EmissionModel, sites: GaussianSites,
+                 time_grid_shape: tf.TensorShape, observations_index: tf.Tensor, observations: tf.Tensor):
+        """
+        :param state_space_model: Parameterises the latent chain.
+        :param emission_model: Maps the latent chain to the observations.
+        :param sites: Gaussian sites over the time grid.
+        :param time_grid_shape: Time grid shape, tuple (T, 1).
+        :param observations_index: Index of the observations in the time grid with shape (N,).
+        :param observations: Observations with shape (N, state_dim).
```
Comment (vincentadam87) [inline markovflow/kalman_filter.py]: observations have size output_dim

File: markovflow/kalman_filter.py [hunk 8]
```diff
@@ -495,3 +494,119 @@ def _log_det_observation_precision(self):
...
+        """
+        self.sites = sites
+        self.time_grid_shape = time_grid_shape
+        self.observations_index = observations_index
+        self._observations = tf.scatter_nd(self.observations_index, observations, self.time_grid_shape)
+        self.data_sites = self._get_data_sites()
+        super().__init__(state_space_model, emission_model)
```
Comment (vincentadam87) [inline markovflow/kalman_filter.py]:
```suggestion
        self.sparse_observations = observations
        self.data_sites = self._get_data_sites()
        super().__init__(state_space_model, emission_model)
        
    @property
    def _observations(self):
        """
        The sparse vector of observation of size on the time grid.
        """
        return tf.scatter_nd(self.observations_index, self.sparse_observations, self.time_grid_shape)

```

Comment (vincentadam87) [reply markovflow/kalman_filter.py]: I think it is best to save the sparse vector and to have a method to expand it

then you don't need the gather_nd later on

File: markovflow/kalman_filter.py [hunk 9]
```diff
@@ -495,3 +494,119 @@ def _log_det_observation_precision(self):
...
+    and :class:`~markovflow.emission_model.EmissionModel`, with Gaussian sites, over a time grid.
+    """
+
+    def __init__(self, state_space_model: StateSpaceModel, emission_model: EmissionModel, sites: GaussianSites,
+                 time_grid_shape: tf.TensorShape, observations_index: tf.Tensor, observations: tf.Tensor):
+        """
+        :param state_space_model: Parameterises the latent chain.
+        :param emission_model: Maps the latent chain to the observations.
+        :param sites: Gaussian sites over the time grid.
+        :param time_grid_shape: Time grid shape, tuple (T, 1).
```
Comment (vincentadam87) [inline markovflow/kalman_filter.py]: If it is just (T,1) , maybe simpler to same the num_grid_points as an integer.

you can build the shape in the code -> shape = (num_grid_points, 1)
That makes the interface simpler.

File: markovflow/kalman_filter.py [hunk 10]
```diff
@@ -495,3 +494,119 @@ def _log_det_observation_precision(self):
...
+        """
+        return self.data_sites.precisions
+
+    def log_likelihood(self) -> tf.Tensor:
+        r"""
+        Construct a TensorFlow function to compute the likelihood.
+
+        Here, as it is for sparse sites, the num_data is set to number of observations and calculate disp_data by
+        gather the data from disp variable using index of observations.
+        For more mathematical details, look at the log_likelihood function of the parent class.
```
Comment (vincentadam87) [inline markovflow/kalman_filter.py]:
```suggestion
        For more mathematical details, look at the log_likelihood function of the parent class.
        The main difference from the parent class are that the vector of observations is now sparse.        
```

File: markovflow/kalman_filter.py [hunk 11]
```diff
@@ -495,3 +494,119 @@ def _log_det_observation_precision(self):
...
+        # K⁻¹ + GᵀΣ⁻¹G = LLᵀ.
+        l_post = self._k_inv_post.cholesky
+        num_data = self.observations_index.shape[0]
+
+        # Hμ [..., num_transitions + 1, output_dim]
+        marginal = self.emission.project_state_to_f(self.prior_ssm.marginal_means)
+
+        # y = obs - Hμ [..., num_transitions + 1, output_dim]
+        disp = self.observations - marginal
+        disp_data = tf.expand_dims(tf.gather_nd(tf.reshape(disp, (-1, 1)), self.observations_index), axis=0)
```
Comment (vincentadam87) [inline markovflow/kalman_filter.py]:
```suggestion
        # [..., num_data, output_dim]
        disp_data = tf.expand_dims(tf.gather_nd(tf.reshape(disp, (-1, 1)), self.observations_index), axis=0)
```

File: markovflow/kalman_filter.py [hunk 12]
```diff
@@ -495,3 +494,119 @@ def _log_det_observation_precision(self):
...
+        nat2 = tf.gather_nd(self.sites.nat2, self.observations_index)
+        log_norm = tf.gather_nd(self.sites.log_norm, self.observations_index)
+
+        return UnivariateGaussianSitesNat(
+            nat1=nat1,
+            nat2=nat2,
+            log_norm=log_norm,
+        )
+
+    @property
```
Comment (vincentadam87) [inline markovflow/kalman_filter.py:522]:
```suggestion
     def sparse_to_dense(self, tensor):
          return tf.scatter_nd(tensor, self.observations_index)

     def dense_to_sparse(self, tensor):
          return tf.gather_nd(tensor, self.observations_index)


    @property
```

Comment (vincentadam87) [reply markovflow/kalman_filter.py:522]: these shouldn't work as is, but I think it would make the code more readable to add and use such functions throughout the code

File: markovflow/kalman_filter.py (hunk 11 above)
Comment (vincentadam87) [inline markovflow/kalman_filter.py]:
```suggestion
        disp_data = self.sparse_observations - self.dense_to_sparse(marginals)
```

File: tests/integration/test_kalman_filter_with_sparse_sites.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            kernel=kernel,
+            mean_function=LinearMeanFunction(1.1),
+            chol_obs_covariance=tf.constant([[np.sqrt(observation_covariance)]], dtype=default_float()),
+        )
+
+    prior_ssm = kernel.state_space_model(time_grid)
+    emission_model = kernel.generate_emission_model(time_grid)
+    observations_index = tf.where(tf.equal(time_grid[..., None], time_points))[:, 0][..., None]
+
+    observations -= gpr_model.mean_function(time_points)
```
Comment (vincentadam87) [inline tests/integration/test_kalman_filter_with_sparse_sites.py:54]: this does not do anything right? there is no mean function

File: tests/integration/test_kalman_filter_with_sparse_sites.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+    nat1 = observations / observation_covariance
+    nat1 = tf.scatter_nd(observations_index, nat1, time_grid[..., None].shape)
+    nat2 = (-0.5 / observation_covariance) * tf.ones_like(nat1)[..., None]
+    lognorm = tf.zeros_like(nat1)
+    sites = UnivariateGaussianSitesNat(nat1=nat1, nat2=nat2, log_norm=lognorm)
```
Comment (vincentadam87) [inline tests/integration/test_kalman_filter_with_sparse_sites.py:59]: I m not sure why you prefer dense sites for this method 
For training purposes that will create bigger parameters with no gradient flowing through these.

I guess I had in mind to then refactor the SDESSM class you have to use the Sparse CVI so maybe the dense sites make more sense.

**Review by vincentadam87 — COMMENTED** [review]
Second batch of comments.
Feel free to disagree

File: markovflow/kalman_filter.py (hunk 8 above)
Comment (prakharverma) [reply markovflow/kalman_filter.py]: Currently, how it is done is that we calculate `disp = observations - marginals` and then gather on these `disp` so saving sparse_observations won't to much as then we will need to do gather on marginals anyways.

File: tests/integration/test_kalman_filter_with_sparse_sites.py (hunk 13 above)
Comment (prakharverma) [reply tests/integration/test_kalman_filter_with_sparse_sites.py:54]: There is when we initialize the GPR model with 'mean_function=LinearMeanFunction(1.1)'.

File: markovflow/kalman_filter.py (hunk 8 above)
Comment (prakharverma) [reply markovflow/kalman_filter.py]: I see what you mean here now.

Comment (prakharverma) [reply markovflow/kalman_filter.py]: Incorporated it this way, the benefit is that we only save sparse observations but for 'disp' and `disp_data` we do sparse_to_dense and vice versa.

File: tests/integration/test_kalman_filter_with_sparse_sites.py (hunk 14 above)
Comment (prakharverma) [reply tests/integration/test_kalman_filter_with_sparse_sites.py:59]: I see what you mean. So, posterior is only dependent on the data-sites so it can be done this way. 

My assumption was that all-sites should also contribute to the posterior which it doesn't. Shouldn't it also play a role? Otherwise, during inference what are the all-sites used for?

Comment (vincentadam87) [reply tests/integration/test_kalman_filter_with_sparse_sites.py:59]: the assumption of this model is that the all sites are forever 0

**Review by vincentadam87 — APPROVED** [review]
Looks good

**Review by vincentadam87 — APPROVED** [review]
Looks good!
